### PR TITLE
Convert semaphore slim to tcs to reduce flakiness

### PR DESCRIPTION
**Pull request by JunTaoLuo**
This should help with https://github.com/aspnet/AspNetCore-Internal/issues/1326.

cc @muratg for 2.1 approval. This is a test only change.

### CODE REVIEW

**Comment by muratg** [comment]
Approved for 2.1 test-only check-in.

File: src/Caching/Memory/test/Infrastructure/TaskExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Caching.Memory.Tests.Infrastructure
+{
+    public static class TaskExtensions
+    {
+        public static async Task<bool> WaitAsync(this Task task, TimeSpan timeout)
```
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply src/Caching/Memory/test/Infrastructure/TaskExtensions.cs]: Sure.

[the remark this answers is not in the document]
Comment (Tratcher) [reply src/Caching/Memory/test/Infrastructure/TaskExtensions.cs]: TimeoutAfter

File: src/Caching/Memory/test/CapacityTests.cs [hunk 2]
```diff
@@ -155,10 +157,10 @@ public async Task ExceedsCapacityCompacts()
             });
 
             var entryOptions = new MemoryCacheEntryOptions { Size = 6 };
-            var sem = new SemaphoreSlim(0, 1);
+            var tcs = new TaskCompletionSource<object>();
```
Comment (Tratcher) [inline src/Caching/Memory/test/CapacityTests.cs]: TCS needs TaskCreationOptions.RunContinuationsAsynchronously to actually help here.

Comment (JunTaoLuo) [reply src/Caching/Memory/test/CapacityTests.cs]: Ah, now I recall. I was looking into this test failure before my vacation and it seemed more like a threadpool starvation issue now that more tests are run in parallel. I added logging to confirm this but I just realized the build wasn't capturing the functional logs. I've fixed that in https://github.com/aspnet/Extensions/pull/799. I'll wait for the logs to confirm.

**Comment by JunTaoLuo** [comment]
I'm going to close this PR. I still haven't seen any functional logs to confirm whether this is a threadpool starvation issue. It seems like this issue is rare as it hasn't occured in almost 3 weeks.

**Pull request closed without merging**